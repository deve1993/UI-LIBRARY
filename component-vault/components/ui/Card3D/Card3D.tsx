import React, { useRef, useState } from 'react';
import { cn } from '@/shared/utils/cn';
import type { Card3DProps } from './Card3D.types';

/**
 * Card3D Component
 *
 * Interactive card with 3D tilt effect on mouse move.
 * Creates an immersive, premium feel for showcasing content.
 *
 * @example
 * ```tsx
 * <Card3D>
 *   <div className="p-8">
 *     <h3 className="text-2xl font-bold">Card Title</h3>
 *     <p>Card content...</p>
 *   </div>
 * </Card3D>
 * ```
 */
export const Card3D: React.FC<Card3DProps> = ({
  children,
  className,
  containerClassName,
}) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const [rotateX, setRotateX] = useState(0);
  const [rotateY, setRotateY] = useState(0);

  const handleMouseMove = (e: React.MouseEvent<HTMLDivElement>) => {
    if (!containerRef.current) return;

    const rect = containerRef.current.getBoundingClientRect();
    const x = e.clientX - rect.left;
    const y = e.clientY - rect.top;

    const centerX = rect.width / 2;
    const centerY = rect.height / 2;

    const rotateXValue = ((y - centerY) / centerY) * -10;
    const rotateYValue = ((x - centerX) / centerX) * 10;

    setRotateX(rotateXValue);
    setRotateY(rotateYValue);
  };

  const handleMouseLeave = () => {
    setRotateX(0);
    setRotateY(0);
  };

  return (
    <div
      ref={containerRef}
      onMouseMove={handleMouseMove}
      onMouseLeave={handleMouseLeave}
      className={cn('perspective-1000', containerClassName)}
      style={{ perspective: '1000px' }}
    >
      <div
        className={cn(
          'relative transition-transform duration-300 ease-out',
          'transform-gpu',
          className
        )}
        style={{
          transform: `rotateX(${rotateX}deg) rotateY(${rotateY}deg)`,
          transformStyle: 'preserve-3d',
        }}
      >
        {children}
      </div>
    </div>
  );
};
