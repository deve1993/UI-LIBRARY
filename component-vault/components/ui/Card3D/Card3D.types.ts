import type { ReactNode } from 'react';

export interface Card3DProps {
  /**
   * Card content
   */
  children: ReactNode;

  /**
   * Additional CSS classes
   */
  className?: string;

  /**
   * Container CSS classes
   */
  containerClassName?: string;
}
