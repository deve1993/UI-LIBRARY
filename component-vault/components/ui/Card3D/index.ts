export { Card3D } from './Card3D';
export type { Card3DProps } from './Card3D.types';
