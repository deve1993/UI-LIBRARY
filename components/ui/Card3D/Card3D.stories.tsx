import type { Meta, StoryObj } from '@storybook/react';
import { Card3D } from './Card3D';
import { Sparkles, Zap, Shield } from 'lucide-react';

const meta: Meta<typeof Card3D> = {
  title: 'UI/Card3D',
  component: Card3D,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
  },
};

export default meta;
type Story = StoryObj<typeof Card3D>;

export const Default: Story = {
  render: () => (
    <Card3D className="bg-white rounded-xl shadow-2xl p-8 w-80">
      <h3 className="text-2xl font-bold mb-4 text-gray-900">
        Card 3D Effect
      </h3>
      <p className="text-gray-600">
        Muovi il mouse sopra questa card per vedere l'effetto 3D interattivo!
      </p>
    </Card3D>
  ),
};

export const GradientCard: Story = {
  render: () => (
    <Card3D className="bg-gradient-to-br from-blue-500 to-purple-600 rounded-xl shadow-2xl p-8 w-80">
      <div className="flex items-center gap-3 mb-4">
        <Sparkles className="w-8 h-8 text-white" />
        <h3 className="text-2xl font-bold text-white">Premium Feature</h3>
      </div>
      <p className="text-white/90">
        Effetto tilt 3D con gradiente colorato per un look premium
      </p>
    </Card3D>
  ),
};

export const FeatureCards: Story = {
  render: () => (
    <div className="grid grid-cols-1 md:grid-cols-3 gap-6 p-8">
      <Card3D className="bg-white rounded-xl shadow-lg p-6">
        <div className="flex flex-col items-center text-center">
          <div className="w-12 h-12 bg-blue-100 rounded-full flex items-center justify-center mb-4">
            <Zap className="w-6 h-6 text-blue-600" />
          </div>
          <h3 className="text-xl font-bold mb-2">Velocità</h3>
          <p className="text-gray-600">Performance ottimizzate per un'esperienza fluida</p>
        </div>
      </Card3D>

      <Card3D className="bg-white rounded-xl shadow-lg p-6">
        <div className="flex flex-col items-center text-center">
          <div className="w-12 h-12 bg-purple-100 rounded-full flex items-center justify-center mb-4">
            <Sparkles className="w-6 h-6 text-purple-600" />
          </div>
          <h3 className="text-xl font-bold mb-2">Design</h3>
          <p className="text-gray-600">Interfaccia moderna e intuitiva</p>
        </div>
      </Card3D>

      <Card3D className="bg-white rounded-xl shadow-lg p-6">
        <div className="flex flex-col items-center text-center">
          <div className="w-12 h-12 bg-green-100 rounded-full flex items-center justify-center mb-4">
            <Shield className="w-6 h-6 text-green-600" />
          </div>
          <h3 className="text-xl font-bold mb-2">Sicurezza</h3>
          <p className="text-gray-600">Protezione avanzata dei tuoi dati</p>
        </div>
      </Card3D>
    </div>
  ),
};

export const DarkCard: Story = {
  render: () => (
    <div className="bg-slate-900 p-12">
      <Card3D className="bg-slate-800 border border-slate-700 rounded-xl shadow-2xl p-8 w-80">
        <h3 className="text-2xl font-bold mb-4 text-white">
          Dark Mode
        </h3>
        <p className="text-slate-300">
          Perfetto anche in modalità scura con bordi sottili
        </p>
      </Card3D>
    </div>
  ),
};
