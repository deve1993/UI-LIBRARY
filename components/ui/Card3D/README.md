# Card3D

Interactive card component with 3D tilt effect on mouse movement.

## Features

- üñ±Ô∏è Interactive 3D tilt effect
- ‚ö° Smooth animations
- üé® Fully customizable styling
- üì± Touch-friendly
- üéØ Perfect for showcasing features, products, or content

## Usage

```tsx
import { Card3D } from '@/components/ui/Card3D';

function FeatureCard() {
  return (
    <Card3D className="bg-white rounded-xl shadow-xl p-8">
      <h3 className="text-2xl font-bold mb-4">Amazing Feature</h3>
      <p className="text-gray-600">
        This card tilts in 3D as you move your mouse over it!
      </p>
    </Card3D>
  );
}
```

## Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `children` | `ReactNode` | - | Card content |
| `className` | `string` | - | Additional CSS classes for the card |
| `containerClassName` | `string` | - | Additional CSS classes for the container |

## Examples

### Basic Card

```tsx
<Card3D className="bg-white p-6 rounded-lg shadow-lg">
  <h3>Card Title</h3>
  <p>Card content</p>
</Card3D>
```

### Feature Card

```tsx
<Card3D className="bg-gradient-to-br from-blue-500 to-purple-600 p-8 rounded-xl text-white">
  <div className="flex items-center gap-4 mb-4">
    <Icon className="w-12 h-12" />
    <h3 className="text-2xl font-bold">Premium Feature</h3>
  </div>
  <p className="text-white/90">
    Experience the next level of interaction
  </p>
</Card3D>
```

### Grid of Cards

```tsx
<div className="grid grid-cols-1 md:grid-cols-3 gap-6">
  {features.map((feature) => (
    <Card3D key={feature.id} className="bg-white p-6 rounded-xl">
      <h3>{feature.title}</h3>
      <p>{feature.description}</p>
    </Card3D>
  ))}
</div>
```

## How it Works

The component uses CSS 3D transforms and React state to create the tilt effect:

1. Tracks mouse position relative to the card
2. Calculates rotation angles based on mouse position
3. Applies smooth transforms using CSS transitions
4. Resets to original position on mouse leave

## Source

Adapted from [Aceternity UI](https://ui.aceternity.com/components/3d-card-effect)
