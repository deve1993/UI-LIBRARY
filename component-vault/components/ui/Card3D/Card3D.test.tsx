import { describe, it, expect, vi } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { axe, toHaveNoViolations } from 'jest-axe'
import { Card3D } from './Card3D'
import type { Card3DProps } from './Card3D.types'

expect.extend(toHaveNoViolations)

describe('Card3D', () => {
  const defaultProps: Card3DProps = {
    children: 'Test Content',
  }

  describe('Rendering', () => {
    it('renders without crashing', () => {
      render(<Card3D {...defaultProps} />)
      expect(screen.getByTestId('card3d')).toBeInTheDocument()
    })

    it('renders children correctly', () => {
      const testContent = 'Test Content'
      render(<Card3D {...defaultProps}>{testContent}</Card3D>)
      expect(screen.getByText(testContent)).toBeInTheDocument()
    })
  })

  describe('Props', () => {

    it('applies className prop correctly', () => {
      const testValue = 'Test'
      render(<Card3D {...defaultProps} className={testValue} />)
      const element = screen.getByTestId('card3d')
      expect(element).toBeInTheDocument()
      // Add specific assertion based on prop type
    })

    it('applies containerClassName prop correctly', () => {
      const testValue = 'Test'
      render(<Card3D {...defaultProps} containerClassName={testValue} />)
      const element = screen.getByTestId('card3d')
      expect(element).toBeInTheDocument()
      // Add specific assertion based on prop type
    })
  })

  describe('Interactions', () => {

    it('handles onMouseMove correctly', async () => {
      const handler = vi.fn()
      const user = userEvent.setup()
      render(<Card3D {...defaultProps} onMouseMove={handler} />)

      const element = screen.getByTestId('card3d')
      await user.click(element)

      expect(handler).toHaveBeenCalledTimes(1)
    })

    it('handles onMouseLeave correctly', async () => {
      const handler = vi.fn()
      const user = userEvent.setup()
      render(<Card3D {...defaultProps} onMouseLeave={handler} />)

      const element = screen.getByTestId('card3d')
      await user.click(element)

      expect(handler).toHaveBeenCalledTimes(1)
    })
  })

  describe('Accessibility', () => {
    it('has no accessibility violations', async () => {
      const { container } = render(<Card3D {...defaultProps} />)
      const results = await axe(container)
      expect(results).toHaveNoViolations()
    })

    it('supports keyboard navigation', async () => {
      const user = userEvent.setup()
      render(<Card3D {...defaultProps} />)
      await user.tab()
      const focusedElement = screen.getByTestId('card3d')
      expect(document.activeElement).toBe(focusedElement)
    })

    it('has correct ARIA attributes', () => {
      render(<Card3D {...defaultProps} />)
      const element = screen.getByTestId('card3d')
      expect(element).toHaveAttribute('role')
    })
  })

  describe('Visual Regression', () => {
    it('matches snapshot', () => {
      const { container } = render(<Card3D {...defaultProps} />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
