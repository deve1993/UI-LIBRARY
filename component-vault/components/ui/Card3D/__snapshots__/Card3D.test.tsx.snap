// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card3D > Visual Regression > matches snapshot 1`] = `
<div
  class="perspective-1000"
  style="perspective: 1000px;"
>
  <div
    class="relative transition-transform duration-300 ease-out transform-gpu"
    style="transform: rotateX(0deg) rotateY(0deg); transform-style: preserve-3d;"
  >
    Test Content
  </div>
</div>
`;
